src/connectors/source.rs

```diff
@@ -0,0 +1,131 @@
+// Copyright 2021, The Tremor Team
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use crate::errors::{Error, Result};
+use crate::pipeline;
+use crate::url::TremorUrl;
+use async_channel::{bounded, Receiver, Sender};
+use beef::Cow;
+use halfbrown::HashMap;
+
+pub(crate) enum SourceMsg {
+    Connect {
+        port: Cow<'static, str>,
+        pipelines: Vec<(TremorUrl, pipeline::Addr)>,
+    },
+    Disconnect {
+        port: Cow<'static, str>,
+        id: TremorUrl,
+    },
+    // TODO: fill those
+    Start,
+    Pause,
+    Resume,
+    Stop,
+}
+
+/// source part of a connector
+#[async_trait::async_trait]
+pub(crate) trait Source: Send {
+    /// Pulls an event from the source if one exists
+    /// determine the codec to be used
+    async fn pull_data(&mut self, id: u64) -> Result<crate::source::SourceReply>;
+    /// This callback is called when the data provided from
+    /// pull_event did not create any events, this is needed for
+    /// linked sources that require a 1:1 mapping between requests
+    /// and responses, we're looking at you REST
+    async fn on_no_events(&mut self, _id: u64, _stream: usize) -> Result<()> {
+        Ok(())
+    }
+    // TODO: lifecycle callbacks
+}
+
+pub(crate) struct ChannelSource {
+    rx: Receiver<crate::source::SourceReply>,
+    tx: Sender<crate::source::SourceReply>,
+}
+
+impl ChannelSource {
+    pub(crate) fn new(qsize: usize) -> Self {
+        let (tx, rx) = bounded(qsize);
+        Self { tx, rx }
+    }
+
+    pub(crate) fn sender(&self) -> Sender<crate::source::SourceReply> {
+        self.tx.clone()
+    }
+}
+
+#[async_trait::async_trait()]
+impl Source for ChannelSource {
+    async fn pull_data(&mut self, _id: u64) -> Result<crate::source::SourceReply> {
+        if self.rx.is_empty() {
```

Comment:
https://docs.rs/async-channel/1.6.1/async_channel/struct.Receiver.html#method.try_recv might be more applicable here?

Reply:
i had it here, i though, maybe checking a number on the queue might be faster 😎